File: 프로그래머스/LV2/코딩테스트 연습/2022 KAKAO BLIND RECRUITMENT/양궁대회/양궁대회.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> ans(11, 0);
int max_diff = 0;

bool check_low_score(vector<int>& lion) {
    for(int i = 10; i >= 0; i--) {
        if(lion[i] > ans[i])return true;
        else if(lion[i] < ans[i])return false;
    }
}

void go(int idx, int n, vector<int>& apeach, vector<int>& lion) {
    if(idx == 11 || n < 0)return;

    if(idx == 10 && n >= 0) {
        lion[idx] = n;

        int lion_score = 0;
        int apeach_score = 0;

        for(int i = 0; i <= 10; i++) {
            if(lion[i] > apeach[i]) lion_score += 10-i;
            else if(apeach[i] != 0) apeach_score += 10-i;
        }

        int diff = lion_score - apeach_score;
        if(diff > max_diff) {
            max_diff = diff;
            copy(lion.begin(), lion.end(), ans.begin());
        }
        else if(diff == max_diff) {
            if(check_low_score(lion))
                copy(lion.begin(), lion.end(), ans.begin());
        }
        return;
    }

    lion[idx] += apeach[idx]+1;
    go(idx+1, n-(apeach[idx]+1), apeach, lion);
    lion[idx] = 0;
    go(idx+1, n, apeach, lion);
}

vector<int> solution(int n, vector<int> info) {
    vector<int> lion(11,0);
    go(0, n, info, lion);
    if(!max_diff)ans = {-1};
    return ans;
}

File: 프로그래머스/LV2/코딩테스트 연습/2022 KAKAO BLIND RECRUITMENT/양궁대회/양궁대회.py
ans = [0]*11
max_diff = 0

def check_low_score(lion):
    for i in range(10, -1, -1):
        if lion[i] > ans[i]:return True
        elif lion[i] < ans[i]:return False

def go(j, cnt, n, apeach, lion):
    global ans, max_diff
    
    if cnt == n:
        lion_score, apeach_score = 0, 0

        for i in range(11):
            if lion[i] > apeach[i]:lion_score += 10-i
            elif apeach[i] != 0:apeach_score += 10-i

        diff = lion_score-apeach_score
        if diff > max_diff:
            max_diff = diff
            ans = lion.copy()
        elif diff == max_diff:
            if check_low_score(lion):
                ans = lion.copy()
        return

    for i in range(j, 11):
        lion[i] += 1
        if lion[i] > apeach[i]:
            go(i+1, sum(lion), n, apeach, lion)
        else:
            go(i, sum(lion), n, apeach, lion)
        lion[i] = 0

def solution(n, info):
    global ans, max_diff
    for i in range(11):
        go(i, 0, n, info, [0]*11)
    if max_diff == 0:ans = [-1]
    return ans